[flaky unit test] in pkg/controller/admissionchecks/multikueue

**Which test is flaking?**:

unit test in sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue

**Link to failed CI job or steps to reproduce locally**:

https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/kubernetes-sigs_kueue/9433/pull-kueue-test-unit-main/2026291415419457536
**Failure message or logs**:
```
=== Failed
=== FAIL: pkg/controller/admissionchecks/multikueue  (0.00s)
Goroutine 19312 (running) created at:
  sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue.(*KubeConfigFSWatcher).Start()
      /home/prow/go/src/sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue/fswatch.go:70 +0x244
  sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue.TestFSWatchAddRm.func2()
      /home/prow/go/src/sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue/fswatch_test.go:228 +0x44
  sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue.TestFSWatchAddRm.func13()
      /home/prow/go/src/sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue/fswatch_test.go:443 +0x5a
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1934 +0x21c
  testing.(*T).Run.gowrap1()
      /usr/local/go/src/testing/testing.go:1997 +0x44
Goroutine 19306 (finished) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:1997 +0x9d2
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:2477 +0x85
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:1934 +0x21c
  testing.runTests()
      /usr/local/go/src/testing/testing.go:2475 +0x96c
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:2337 +0xed4
  main.main()
      _testmain.go:81 +0x164
==================
FAIL
coverage: 5.0% of statements in ./...
FAIL	sigs.k8s.io/kueue/pkg/controller/admissionchecks/multikueue	85.035s
DONE 12875 tests, 9 skipped, 1 failure in 484.947s
make: *** [Makefile-test.mk:77: test] Error 1
```

**Anything else we need to know?**:

--- a/pkg/controller/admissionchecks/multikueue/fswatch.go
+++ b/pkg/controller/admissionchecks/multikueue/fswatch.go
@@ -38,6 +38,7 @@ var (
 
 type KubeConfigFSWatcher struct {
 	watcher       *fsnotify.Watcher
+	ready         chan struct{}
 	lock          sync.RWMutex
 	clusterToFile map[string]string
 	// a single file can be potentially used by multiple clusters
@@ -50,6 +51,7 @@ var _ manager.Runnable = (*KubeConfigFSWatcher)(nil)
 
 func newKubeConfigFSWatcher() *KubeConfigFSWatcher {
 	return &KubeConfigFSWatcher{
+		ready:            make(chan struct{}),
 		clusterToFile:    map[string]string{},
 		fileToClusters:   map[string]set.Set[string]{},
 		parentDirToFiles: map[string]set.Set[string]{},
@@ -59,42 +61,39 @@ func newKubeConfigFSWatcher() *KubeConfigFSWatcher {
 
 func (w *KubeConfigFSWatcher) Start(ctx context.Context) error {
 	w.lock.Lock()
-	defer w.lock.Unlock()
-
 	var err error
 	w.watcher, err = fsnotify.NewWatcher()
+	w.lock.Unlock()
 	if err != nil {
 		return err
 	}
+	close(w.ready)
 
-	go func() {
-		log := ctrl.LoggerFrom(ctx).WithName("kubeconfig-fs-watcher")
-		defer func() {
-			err := w.watcher.Close()
-			if err != nil {
-				log.Error(err, "Closing kubeconfigs FS Watcher")
-			}
-		}()
-		watchedEvents := fsnotify.Write | fsnotify.Create | fsnotify.Remove | fsnotify.Rename
-		for {
-			select {
-			case ev := <-w.watcher.Events:
-				log.V(5).Info("Got event", "name", ev.Name, "op", ev.Op)
-				switch {
-				case (ev.Op & watchedEvents) != 0:
-					w.notifyPathWrite(ev.Name)
-				default:
-					log.V(2).Info("Ignore event", "name", ev.Name, "op", ev.Op)
-				}
-			case err := <-w.watcher.Errors:
-				log.Error(err, "Kubeconfigs FS Watch")
-			case <-ctx.Done():
-				log.V(2).Info("End kubeconfigs FS Watch")
-				return
-			}
+	log := ctrl.LoggerFrom(ctx).WithName("kubeconfig-fs-watcher")
+	defer func() {
+		err := w.watcher.Close()
+		if err != nil {
+			log.Error(err, "Closing kubeconfigs FS Watcher")
 		}
 	}()
-	return nil
+	watchedEvents := fsnotify.Write | fsnotify.Create | fsnotify.Remove | fsnotify.Rename
+	for {
+		select {
+		case ev := <-w.watcher.Events:
+			log.V(5).Info("Got event", "name", ev.Name, "op", ev.Op)
+			switch {
+			case (ev.Op & watchedEvents) != 0:
+				w.notifyPathWrite(ctx, ev.Name)
+			default:
+				log.V(2).Info("Ignore event", "name", ev.Name, "op", ev.Op)
+			}
+		case err := <-w.watcher.Errors:
+			log.Error(err, "Kubeconfigs FS Watch")
+		case <-ctx.Done():
+			log.V(2).Info("End kubeconfigs FS Watch")
+			return nil
+		}
+	}
 }
 
 func (w *KubeConfigFSWatcher) clustersForPath(path string) []string {
@@ -103,9 +102,13 @@ func (w *KubeConfigFSWatcher) clustersForPath(path string) []string {
 	return w.fileToClusters[path].UnsortedList()
 }
 
-func (w *KubeConfigFSWatcher) notifyPathWrite(path string) {
+func (w *KubeConfigFSWatcher) notifyPathWrite(ctx context.Context, path string) {
 	for _, c := range w.clustersForPath(path) {
-		w.reconcile <- event.GenericEvent{Object: &kueue.MultiKueueCluster{ObjectMeta: metav1.ObjectMeta{Name: c}}}
+		select {
+		case w.reconcile <- event.GenericEvent{Object: &kueue.MultiKueueCluster{ObjectMeta: metav1.ObjectMeta{Name: c}}}:
+		case <-ctx.Done():
+			return
+		}
 	}
 }
 
@@ -186,9 +189,12 @@ func (w *KubeConfigFSWatcher) cleanOldWatchDirs() error {
 }
 
 func (w *KubeConfigFSWatcher) Started() bool {
-	w.lock.RLock()
-	defer w.lock.RUnlock()
-	return w.watcher != nil
+	select {
+	case <-w.ready:
+		return true
+	default:
+		return false
+	}
 }
 
 func (w *KubeConfigFSWatcher) AddOrUpdate(cluster, path string) error {

--- a/pkg/controller/admissionchecks/multikueue/fswatch_test.go
+++ b/pkg/controller/admissionchecks/multikueue/fswatch_test.go
@@ -149,12 +149,15 @@ func TestFSWatch(t *testing.T) {
 			// Using an empty context here to avoid a race condition with the test context when setting the logger.
 			ctx := ctrl.LoggerInto(context.Background(), utiltesting.NewLogger(t))
 			ctx, cancel := context.WithTimeout(ctx, time.Second)
-			defer cancel()
+			wg := sync.WaitGroup{}
+			defer func() {
+				cancel()
+				wg.Wait()
+			}()
 			watcher := newKubeConfigFSWatcher()
 
-			// start the recorder
+			// start the recorder and the watcher
 			gotEventsForClusters := set.New[string]()
-			wg := sync.WaitGroup{}
 			wg.Go(func() {
 				for {
 					select {
@@ -168,7 +171,10 @@ func TestFSWatch(t *testing.T) {
 					}
 				}
 			})
-			_ = watcher.Start(ctx)
+			wg.Go(func() {
+				_ = watcher.Start(ctx)
+			})
+			<-watcher.ready
 
 			gotAddErrors := map[string]error{}
 			for c, p := range tc.clusters {
@@ -200,7 +206,11 @@ func TestFSWatchAddRm(t *testing.T) {
 	// Using an empty context here to avoid a race condition with the test context when setting the logger.
 	ctx := ctrl.LoggerInto(context.Background(), utiltesting.NewLogger(t))
 	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
+	wg := sync.WaitGroup{}
+	defer func() {
+		cancel()
+		wg.Wait()
+	}()
 	basePath := t.TempDir()
 	f1Path := filepath.Join(basePath, "file.one")
 	f2Path := filepath.Join(basePath, "file.two")
@@ -225,7 +235,11 @@ func TestFSWatchAddRm(t *testing.T) {
 		{
 			name: "start",
 			opFnc: func(kcf *KubeConfigFSWatcher) error {
-				return kcf.Start(ctx)
+				wg.Go(func() {
+					_ = kcf.Start(ctx)
+				})
+				<-kcf.ready
+				return nil
 			},
 			wantClustersToFile: map[string]string{},
 			wantFileToClusters: map[string]set.Set[string]{},

